src/zim_types.h

```diff
@@ -31,84 +31,50 @@
   { return v == rhs.v; }
 
-  inline REAL_TYPEDEF<B>& operator++()
+  inline REAL_TYPEDEF<B, S>& operator++()
   { v++; return *this; }
 
-  inline REAL_TYPEDEF<B> operator++(int)
-  { return REAL_TYPEDEF<B>(v++); }
+  inline REAL_TYPEDEF<B, S> operator++(int)
+  { return REAL_TYPEDEF<B, S>(v++); }
 } PACKED;
 
-template<typename T> inline T& operator+= (T& lhs, const T& rhs)
-{
-  lhs.v += rhs.v;
-  return lhs;
-}
-
-template<typename T> inline T& operator+= (T& lhs, const typename T::base_type& rhs)
-{
-  lhs.v += rhs;
-  return lhs;
-}
-
-template<typename T> inline T operator+(T lhs, const T& rhs)
-{
-  lhs += rhs;
-  return lhs;
-}
-
-template<typename T> inline T& operator-=(T& lhs, const T& rhs)
-{
-  lhs.v -= rhs.v;
-  return lhs;
-}
-
-template<typename T> inline T operator-(T lhs, const T& rhs)
-{
-  lhs -= rhs;
-  return lhs;
-}
-
-template<typename T> inline bool operator< (const T& lhs, const T& rhs)
-{ return lhs.v < rhs.v; }
-
-template<typename T> inline bool operator> (const T& lhs, const T& rhs)
-{ return rhs < lhs; }
-
-template<typename T> inline bool operator<=(const T& lhs, const T& rhs)
-{ return !(lhs > rhs); }
-
-template<typename T> inline bool operator>=(const T& lhs, const T& rhs)
-{ return !(lhs < rhs); }
-
-template<typename T> inline bool operator!=(const T& lhs, const T& rhs)
-{ return !(lhs == rhs); }
-
-
-template<typename B>
-std::ostream& operator<<(std::ostream& os, const REAL_TYPEDEF<B>& obj)
+template<typename B, typename S>
+std::ostream& operator<<(std::ostream& os, const REAL_TYPEDEF<B, S>& obj)
 {
     os << obj.v;
     return os;
 }
 
 namespace zim {
 
-#define TYPEDEF(NAME, TYPE) struct NAME : public REAL_TYPEDEF<TYPE> { \
-explicit NAME(TYPE v=0) : REAL_TYPEDEF<TYPE>(v) {}; } PACKED; \
-static_assert(sizeof(NAME) == sizeof(TYPE), "");
+#define TYPEDEF(NAME, TYPE) struct NAME : public REAL_TYPEDEF<TYPE, NAME> {          \
+explicit NAME(TYPE v=0) : REAL_TYPEDEF<TYPE, NAME>(v) {}; } PACKED;                  \
+static_assert(sizeof(NAME) == sizeof(TYPE), "");                                     \
+inline NAME& operator+= (NAME& lhs, const NAME& rhs) { lhs.v += rhs.v; return lhs; } \
+inline NAME& operator+= (NAME& lhs, const TYPE& rhs) { lhs.v += rhs; return lhs; }   \
+inline NAME operator+(NAME lhs, const NAME& rhs) { lhs += rhs; return lhs; }         \
+inline NAME& operator-=(NAME& lhs, const NAME& rhs) { lhs.v -= rhs.v; return lhs; }  \
+inline NAME operator-(NAME lhs, const NAME& rhs) { lhs -= rhs; return lhs; }         \
+inline bool operator< (const NAME& lhs, const NAME& rhs) { return lhs.v < rhs.v; }   \
+inline bool operator> (const NAME& lhs, const NAME& rhs) { return rhs < lhs; }       \
+inline bool operator<=(const NAME& lhs, const NAME& rhs) { return !(lhs > rhs); }    \
+inline bool operator>=(const NAME& lhs, const NAME& rhs) { return !(lhs < rhs); }    \
+inline bool operator!=(const NAME& lhs, const NAME& rhs) { return !(lhs == rhs); }
+
```

Comment:
Ideally, utilities like this one should be tested too. However it requires a different type of testing (that supports checking for expected compilation failure).

Reply:
Indeed. But I prefer to move that to another issue (if we do it)